src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs

```diff
@@ -0,0 +1,173 @@
+            IMethodNode methodEntryPointNode = factory.MethodEntrypoint(method, getUnboxingStub);
+            dependencies.Add(new DependencyListEntry(methodEntryPointNode, "Exact method instantiation entry"));
+
+            // Get native layout dependencies for the declaring type
+            dependencies.Add(new DependencyListEntry(factory.NecessaryTypeSymbol(method.OwningType), "Exact method instantiation entry"));
+
+            // Get native layout dependencies for the method instantiation args
+            foreach (var arg in method.Instantiation)
+                dependencies.Add(new DependencyListEntry(factory.NecessaryTypeSymbol(arg), "Exact method instantiation entry"));
+
+            // Get native layout dependencies for the method signature.
+            NativeLayoutMethodNameAndSignatureVertexNode nameAndSig = factory.NativeLayoutMethodNameAndSignatureVertex(method.GetTypicalMethodDefinition());
+            dependencies.Add(new DependencyListEntry(factory.NativeLayoutPlacedSignatureVertex(nameAndSig), "Exact method instantiation entry"));
+
+            return dependencies;
+        }
+
+        public void AddEntryIfEligible(NodeFactory factory, MethodDesc method)
+        {
+            // Check if we already saw this method
+            if (!_exactMethodInstantiationsList.Add(method))
+                return;
+
+            if (!IsMethodEligibleForTracking(method))
+                return;
+
+            bool getUnboxingStub = method.OwningType.IsValueType && !method.Signature.IsStatic;
+            IMethodNode methodEntryPointNode = factory.MethodEntrypoint(method, getUnboxingStub);
+
+            Vertex methodSignature = GetNativeLayoutSignatureForMethodEntry(factory, method);
+            Vertex methodPointer = _nativeWriter.GetUnsignedConstant(_externalReferences.GetIndex(methodEntryPointNode));
+
+            // Make the generic method entry vertex
+
+            Vertex entry = _nativeWriter.GetTuple(methodSignature, methodPointer);
+
+            // Add to the hash table, hashed by the containing type's hashcode
+            uint hashCode = (uint)method.OwningType.GetHashCode();
+            _hashtable.Append(hashCode, _nativeSection.Place(entry));
+        }
+
+        private static bool IsMethodEligibleForTracking(MethodDesc method)
+        {
+            if (method.IsAbstract)
+                return false;
+
+            if (!method.HasInstantiation)
+                return false;
+
+            // This hashtable is only for method instantiations that don't use generic dictionaries,
+            // so check if the given method is shared before proceeding
+            if (method.IsSharedByGenericInstantiations || method.GetCanonMethodTarget(CanonicalFormKind.Specific) != method)
+                return false;
+
+            // Runtime determined methods should never show up here.
+            Debug.Assert(!method.IsRuntimeDeterminedExactMethod);
```

Comment:
Nit: Argument validation usually happens at the beginning of method. I would want the assert to kick in even if this is e.g. a non-generic method.